crates/goose/src/truncate.rs

```diff
@@ -0,0 +1,706 @@
+use crate::message::Message;
+use anyhow::{anyhow, Result};
+use mcp_core::Role;
+use std::collections::HashSet;
+use tracing::debug;
+
+/// Trait representing a truncation strategy
+pub trait TruncationStrategy {
+    /// Determines the indices of messages to remove to fit within the context limit.
+    ///
+    /// - `messages`: The list of messages in the conversation.
+    /// - `token_counts`: A parallel array containing the token count for each message.
+    /// - `context_limit`: The maximum allowed context length in tokens.
+    ///
+    /// Returns a vector of indices to remove.
+    fn determine_indices_to_remove(
+        &self,
+        messages: &[Message],
+        token_counts: &[usize],
+        context_limit: usize,
+    ) -> Result<HashSet<usize>>;
+}
+
+/// Strategy to truncate messages by removing the oldest first
+pub struct OldestFirstTruncation;
+
+impl TruncationStrategy for OldestFirstTruncation {
+    fn determine_indices_to_remove(
+        &self,
+        messages: &[Message],
+        token_counts: &[usize],
+        context_limit: usize,
+    ) -> Result<HashSet<usize>> {
+        let mut indices_to_remove = HashSet::new();
```

Comment:
honestly i might remove the complexity of having tokenizers entirely if this works, so we never rely on what amount to bad estimates here. Maybe we just drop the first 1/4 and rely on recursion to handle if the data shape is particularly skewed? 

We want some fuzziness anways i think, in that its probably better to give it a stable conversation for several turns rather than truncating the smallest bit possible but on every pass